File: src/runtime2/ruby/pytra/py_runtime.rb
# Ruby native backend runtime helpers.

def __pytra_noop(*args)
  _ = args
  nil
end

def __pytra_assert(*args)
  _ = args
  "True"
end

def __pytra_perf_counter
  Process.clock_gettime(Process::CLOCK_MONOTONIC)
end

def __pytra_truthy(v)
  return false if v.nil?
  return v if v == true || v == false
  return v != 0 if v.is_a?(Integer)
  return v != 0.0 if v.is_a?(Float)
  return !v.empty? if v.respond_to?(:empty?)
  true
end

def __pytra_int(v)
  return 0 if v.nil?
  v.to_i
end

def __pytra_float(v)
  return 0.0 if v.nil?
  v.to_f
end

def __pytra_div(a, b)
  lhs = __pytra_float(a)
  rhs = __pytra_float(b)
  raise ZeroDivisionError, "division by zero" if rhs == 0.0
  lhs / rhs
end

def __pytra_str(v)
  return "" if v.nil?
  v.to_s
end

def __pytra_len(v)
  return 0 if v.nil?
  return v.length if v.respond_to?(:length)
  0
end

def __pytra_as_list(v)
  return v if v.is_a?(Array)
  return v.to_a if v.respond_to?(:to_a)
  []
end

def __pytra_as_dict(v)
  return v if v.is_a?(Hash)
  {}
end

def __pytra_bytearray(v = nil)
  return [] if v.nil?
  if v.is_a?(Integer)
    n = v
    n = 0 if n < 0
    return Array.new(n, 0)
  end
  if v.is_a?(String)
    return v.bytes
  end
  src = __pytra_as_list(v)
  out = []
  i = 0
  while i < src.length
    out << (__pytra_int(src[i]) & 255)
    i += 1
  end
  out
end

def __pytra_bytes(v)
  return [] if v.nil?
  return v.bytes if v.is_a?(String)
  src = __pytra_as_list(v)
  out = []
  i = 0
  while i < src.length
    out << (__pytra_int(src[i]) & 255)
    i += 1
  end
  out
end

def __pytra_range(start_v, stop_v, step_v)
  out = []
  step = __pytra_int(step_v)
  return out if step == 0
  i = __pytra_int(start_v)
  stop = __pytra_int(stop_v)
  while ((step >= 0 && i < stop) || (step < 0 && i > stop))
    out << i
    i += step
  end
  out
end

def __pytra_list_comp_range(start_v, stop_v, step_v)
  out = []
  step = __pytra_int(step_v)
  return out if step == 0
  i = __pytra_int(start_v)
  stop = __pytra_int(stop_v)
  while ((step >= 0 && i < stop) || (step < 0 && i > stop))
    out << yield(i)
    i += step
  end
  out
end

def __pytra_enumerate(v)
  src = __pytra_as_list(v)
  out = []
  i = 0
  while i < src.length
    out << [i, src[i]]
    i += 1
  end
  out
end

def __pytra_abs(v)
  x = __pytra_float(v)
  x < 0 ? -x : x
end

def __pytra_get_index(container, index)
  if container.is_a?(Array)
    i = __pytra_int(index)
    i += container.length if i < 0
    return nil if i < 0 || i >= container.length
    return container[i]
  end
  if container.is_a?(Hash)
    return container[index]
  end
  if container.is_a?(String)
    i = __pytra_int(index)
    i += container.length if i < 0
    return "" if i < 0 || i >= container.length
    return container[i] || ""
  end
  nil
end

def __pytra_set_index(container, index, value)
  if container.is_a?(Array)
    i = __pytra_int(index)
    i += container.length if i < 0
    return if i < 0 || i >= container.length
    container[i] = value
    return
  end
  if container.is_a?(Hash)
    container[index] = value
  end
end

def __pytra_slice(container, lower, upper)
  return nil if container.nil?
  lo = __pytra_int(lower)
  hi = __pytra_int(upper)
  container[lo...hi]
end

def __pytra_min(a, b)
  __pytra_float(a) < __pytra_float(b) ? a : b
end

def __pytra_max(a, b)
  __pytra_float(a) > __pytra_float(b) ? a : b
end

def __pytra_isdigit(v)
  s = __pytra_str(v)
  return false if s.empty?
  !!(s =~ /\A[0-9]+\z/)
end

def __pytra_isalpha(v)
  s = __pytra_str(v)
  return false if s.empty?
  !!(s =~ /\A[A-Za-z]+\z/)
end

def __pytra_contains(container, item)
  return false if container.nil?
  return container.key?(item) if container.is_a?(Hash)
  return container.include?(item) if container.is_a?(Array)
  return container.include?(__pytra_str(item)) if container.is_a?(String)
  false
end

def __pytra_print(*args)
  if args.empty?
    puts
    return
  end
  puts(args.map { |x| __pytra_str(x) }.join(" "))
end

def __pytra_u16le(v)
  x = __pytra_int(v) & 0xFFFF
  [x & 0xFF, (x >> 8) & 0xFF]
end

def __pytra_u32be(v)
  x = __pytra_int(v) & 0xFFFFFFFF
  [(x >> 24) & 0xFF, (x >> 16) & 0xFF, (x >> 8) & 0xFF, x & 0xFF]
end

def __pytra_crc32(data)
  bytes = __pytra_bytes(data)
  crc = 0xFFFFFFFF
  poly = 0xEDB88320
  i = 0
  while i < bytes.length
    crc ^= bytes[i]
    bit = 0
    while bit < 8
      if (crc & 1) != 0
        crc = ((crc >> 1) ^ poly) & 0xFFFFFFFF
      else
        crc = (crc >> 1) & 0xFFFFFFFF
      end
      bit += 1
    end
    i += 1
  end
  (crc ^ 0xFFFFFFFF) & 0xFFFFFFFF
end

def __pytra_adler32(data)
  bytes = __pytra_bytes(data)
  mod = 65_521
  s1 = 1
  s2 = 0
  i = 0
  while i < bytes.length
    s1 += bytes[i]
    s1 -= mod if s1 >= mod
    s2 += s1
    s2 %= mod
    i += 1
  end
  ((s2 << 16) | s1) & 0xFFFFFFFF
end

def __pytra_zlib_deflate_store(data)
  bytes = __pytra_bytes(data)
  out = [0x78, 0x01]
  n = bytes.length
  pos = 0
  while pos < n
    remain = n - pos
    chunk_len = remain > 65_535 ? 65_535 : remain
    final = (pos + chunk_len) >= n ? 1 : 0
    out << final
    out.concat(__pytra_u16le(chunk_len))
    out.concat(__pytra_u16le(0xFFFF ^ chunk_len))
    out.concat(bytes[pos, chunk_len] || [])
    pos += chunk_len
  end
  out.concat(__pytra_u32be(__pytra_adler32(bytes)))
  out
end

def __pytra_png_chunk(chunk_type, data)
  chunk_type_bytes = __pytra_bytes(chunk_type)
  payload = __pytra_bytes(data)
  out = []
  out.concat(__pytra_u32be(payload.length))
  out.concat(chunk_type_bytes)
  out.concat(payload)
  out.concat(__pytra_u32be(__pytra_crc32(chunk_type_bytes + payload)))
  out
end

def __pytra_write_bytes(path, bytes)
  File.binwrite(__pytra_str(path), __pytra_bytes(bytes).pack("C*"))
end

def write_rgb_png(path, width, height, pixels)
  raw = __pytra_bytes(pixels)
  w = __pytra_int(width)
  h = __pytra_int(height)
  expected = w * h * 3
  if raw.length != expected
    raise ArgumentError, "pixels length mismatch: got=#{raw.length} expected=#{expected}"
  end

  scanlines = []
  row_bytes = w * 3
  y = 0
  while y < h
    scanlines << 0
    start = y * row_bytes
    scanlines.concat(raw[start, row_bytes] || [])
    y += 1
  end

  ihdr = __pytra_u32be(w) + __pytra_u32be(h) + [8, 2, 0, 0, 0]
  idat = __pytra_zlib_deflate_store(scanlines)
  png = [0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A]
  png.concat(__pytra_png_chunk([0x49, 0x48, 0x44, 0x52], ihdr))
  png.concat(__pytra_png_chunk([0x49, 0x44, 0x41, 0x54], idat))
  png.concat(__pytra_png_chunk([0x49, 0x45, 0x4E, 0x44], []))
  __pytra_write_bytes(path, png)
  nil
end

def __pytra_gif_lzw_encode(data, min_code_size = 8)
  bytes = __pytra_bytes(data)
  return [] if bytes.empty?

  clear_code = 1 << __pytra_int(min_code_size)
  end_code = clear_code + 1
  code_size = __pytra_int(min_code_size) + 1

  out = []
  bit_buffer = 0
  bit_count = 0

  emit_code = lambda do |code, bits|
    bit_buffer |= (__pytra_int(code) << bit_count)
    bit_count += __pytra_int(bits)
    while bit_count >= 8
      out << (bit_buffer & 0xFF)
      bit_buffer >>= 8
      bit_count -= 8
    end
  end

  emit_code.call(clear_code, code_size)
  i = 0
  while i < bytes.length
    emit_code.call(bytes[i], code_size)
    emit_code.call(clear_code, code_size)
    i += 1
  end
  emit_code.call(end_code, code_size)
  out << (bit_buffer & 0xFF) if bit_count > 0
  out
end

def grayscale_palette
  out = []
  i = 0
  while i < 256
    out << i
    out << i
    out << i
    i += 1
  end
  out
end

def save_gif(path, width, height, frames, palette, delay_cs = 4, loop = 0)
  w = __pytra_int(width)
  h = __pytra_int(height)
  frame_pixels = w * h
  palette_bytes = __pytra_bytes(palette)
  if palette_bytes.length != 256 * 3
    raise ArgumentError, "palette must be 256*3 bytes"
  end

  src_frames = __pytra_as_list(frames)
  normalized_frames = []
  i = 0
  while i < src_frames.length
    fr = __pytra_bytes(src_frames[i])
    if fr.length != frame_pixels
      raise ArgumentError, "frame size mismatch"
    end
    normalized_frames << fr
    i += 1
  end

  out = []
  out.concat([0x47, 0x49, 0x46, 0x38, 0x39, 0x61]) # GIF89a
  out.concat(__pytra_u16le(w))
  out.concat(__pytra_u16le(h))
  out << 0xF7
  out << 0
  out << 0
  out.concat(palette_bytes)

  out.concat([0x21, 0xFF, 0x0B])
  out.concat("NETSCAPE2.0".bytes)
  out.concat([0x03, 0x01])
  out.concat(__pytra_u16le(loop))
  out << 0

  i = 0
  while i < normalized_frames.length
    fr = normalized_frames[i]
    out.concat([0x21, 0xF9, 0x04, 0x00])
    out.concat(__pytra_u16le(delay_cs))
    out.concat([0x00, 0x00])

    out << 0x2C
    out.concat(__pytra_u16le(0))
    out.concat(__pytra_u16le(0))
    out.concat(__pytra_u16le(w))
    out.concat(__pytra_u16le(h))
    out << 0

    out << 8
    compressed = __pytra_gif_lzw_encode(fr, 8)
    pos = 0
    while pos < compressed.length
      chunk = compressed[pos, 255] || []
      out << chunk.length
      out.concat(chunk)
      pos += chunk.length
    end
    out << 0
    i += 1
  end

  out << 0x3B
  __pytra_write_bytes(path, out)
  nil
end

File: src/runtime2/swift/pytra/py_runtime.swift
// Swift 実行向け Node.js ランタイム補助。
// source: src/pytra/utils/png.py
// source: src/pytra/utils/gif.py

import Foundation

/// Base64 で埋め込まれた JavaScript ソースコードを一時ファイルへ展開し、node で実行する。
/// - Parameters:
///   - sourceBase64: JavaScript ソースコードの Base64 文字列。
///   - args: JavaScript 側へ渡す引数配列。
/// - Returns:
///   node プロセスの終了コード。失敗時は 1 を返す。
func pytraRunEmbeddedNode(_ sourceBase64: String, _ args: [String]) -> Int32 {
    guard let sourceData = Data(base64Encoded: sourceBase64) else {
        fputs("error: failed to decode embedded JavaScript source\n", stderr)
        return 1
    }

    let tmpDir = URL(fileURLWithPath: NSTemporaryDirectory(), isDirectory: true)
    let fileName = "pytra_embedded_\(UUID().uuidString).js"
    let scriptURL = tmpDir.appendingPathComponent(fileName)

    do {
        try sourceData.write(to: scriptURL)
    } catch {
        fputs("error: failed to write temporary JavaScript file: \(error)\n", stderr)
        return 1
    }

    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
    process.arguments = ["node", scriptURL.path] + args
    process.environment = ProcessInfo.processInfo.environment
    process.standardInput = FileHandle.standardInput
    process.standardOutput = FileHandle.standardOutput
    process.standardError = FileHandle.standardError

    do {
        try process.run()
        process.waitUntilExit()
    } catch {
        fputs("error: failed to launch node: \(error)\n", stderr)
        try? FileManager.default.removeItem(at: scriptURL)
        return 1
    }

    try? FileManager.default.removeItem(at: scriptURL)
    return process.terminationStatus
}


// ---- legacy swift emitter helper compatibility ----
func __pytra_noop(_ args: Any...) {}

func __pytra_any_default() -> Any {
    return Int64(0)
}

func __pytra_assert(_ args: Any...) -> String {
    _ = args
    return "True"
}

func __pytra_perf_counter() -> Double {
    return Date().timeIntervalSince1970
}

func __pytra_truthy(_ v: Any?) -> Bool {
    guard let value = v else { return false }
    if let b = value as? Bool { return b }
    if let i = value as? Int64 { return i != 0 }
    if let i = value as? Int { return i != 0 }
    if let d = value as? Double { return d != 0.0 }
    if let s = value as? String { return s != "" }
    if let a = value as? [Any] { return !a.isEmpty }
    if let m = value as? [AnyHashable: Any] { return !m.isEmpty }
    return true
}

func __pytra_int(_ v: Any?) -> Int64 {
    guard let value = v else { return 0 }
    if let i = value as? Int64 { return i }
    if let i = value as? Int { return Int64(i) }
    if let d = value as? Double { return Int64(d) }
    if let b = value as? Bool { return b ? 1 : 0 }
    if let s = value as? String { return Int64(s) ?? 0 }
    return 0
}

func __pytra_float(_ v: Any?) -> Double {
    guard let value = v else { return 0.0 }
    if let d = value as? Double { return d }
    if let f = value as? Float { return Double(f) }
    if let i = value as? Int64 { return Double(i) }
    if let i = value as? Int { return Double(i) }
    if let b = value as? Bool { return b ? 1.0 : 0.0 }
    if let s = value as? String { return Double(s) ?? 0.0 }
    return 0.0
}

func __pytra_str(_ v: Any?) -> String {
    guard let value = v else { return "" }
    if let s = value as? String { return s }
    return String(describing: value)
}

func __pytra_len(_ v: Any?) -> Int64 {
    guard let value = v else { return 0 }
    if let s = value as? String { return Int64(s.count) }
    if let a = value as? [Any] { return Int64(a.count) }
    if let m = value as? [AnyHashable: Any] { return Int64(m.count) }
    return 0
}

func __pytra_index(_ i: Int64, _ n: Int64) -> Int64 {
    if i < 0 {
        return i + n
    }
    return i
}

func __pytra_getIndex(_ container: Any?, _ index: Any?) -> Any {
    if let list = container as? [Any] {
        if list.isEmpty { return __pytra_any_default() }
        let i = __pytra_index(__pytra_int(index), Int64(list.count))
        if i < 0 || i >= Int64(list.count) { return __pytra_any_default() }
        return list[Int(i)]
    }
    if let dict = container as? [AnyHashable: Any] {
        let key = AnyHashable(__pytra_str(index))
        return dict[key] ?? __pytra_any_default()
    }
    if let s = container as? String {
        let chars = Array(s)
        if chars.isEmpty { return "" }
        let i = __pytra_index(__pytra_int(index), Int64(chars.count))
        if i < 0 || i >= Int64(chars.count) { return "" }
        return String(chars[Int(i)])
    }
    return __pytra_any_default()
}

func __pytra_setIndex(_ container: Any?, _ index: Any?, _ value: Any?) {
    if var list = container as? [Any] {
        if list.isEmpty { return }
        let i = __pytra_index(__pytra_int(index), Int64(list.count))
        if i < 0 || i >= Int64(list.count) { return }
        list[Int(i)] = value as Any
        return
    }
    if var dict = container as? [AnyHashable: Any] {
        let key = AnyHashable(__pytra_str(index))
        dict[key] = value
    }
}

func __pytra_slice(_ container: Any?, _ lower: Any?, _ upper: Any?) -> Any {
    if let s = container as? String {
        let chars = Array(s)
        let n = Int64(chars.count)
        var lo = __pytra_index(__pytra_int(lower), n)
        var hi = __pytra_index(__pytra_int(upper), n)
        if lo < 0 { lo = 0 }
        if hi < 0 { hi = 0 }
        if lo > n { lo = n }
        if hi > n { hi = n }
        if hi < lo { hi = lo }
        if lo >= hi { return "" }
        return String(chars[Int(lo)..<Int(hi)])
    }
    if let list = container as? [Any] {
        let n = Int64(list.count)
        var lo = __pytra_index(__pytra_int(lower), n)
        var hi = __pytra_index(__pytra_int(upper), n)
        if lo < 0 { lo = 0 }
        if hi < 0 { hi = 0 }
        if lo > n { lo = n }
        if hi > n { hi = n }
        if hi < lo { hi = lo }
        if lo >= hi { return [Any]() }
        return Array(list[Int(lo)..<Int(hi)])
    }
    return __pytra_any_default()
}

func __pytra_isdigit(_ v: Any?) -> Bool {
    let s = __pytra_str(v)
    if s.isEmpty { return false }
    return s.unicodeScalars.allSatisfy { CharacterSet.decimalDigits.contains($0) }
}

func __pytra_isalpha(_ v: Any?) -> Bool {
    let s = __pytra_str(v)
    if s.isEmpty { return false }
    return s.unicodeScalars.allSatisfy { CharacterSet.letters.contains($0) }
}

func __pytra_contains(_ container: Any?, _ value: Any?) -> Bool {
    if let list = container as? [Any] {
        let needle = __pytra_str(value)
        for item in list {
            if __pytra_str(item) == needle {
                return true
            }
        }
        return false
    }
    if let dict = container as? [AnyHashable: Any] {
        return dict[AnyHashable(__pytra_str(value))] != nil
    }
    if let s = container as? String {
        let needle = __pytra_str(value)
        return s.contains(needle)
    }
    return false
}

func __pytra_ifexp(_ cond: Bool, _ a: Any, _ b: Any) -> Any {
    return cond ? a : b
}

func __pytra_to_u8(_ v: Any?) -> UInt8 {
    let n = __pytra_int(v)
    if n < 0 { return 0 }
    if n > 255 { return 255 }
    return UInt8(n)
}

func __pytra_append_u32be(_ out: inout [UInt8], _ value: UInt32) {
    out.append(UInt8((value >> 24) & 0xFF))
    out.append(UInt8((value >> 16) & 0xFF))
    out.append(UInt8((value >> 8) & 0xFF))
    out.append(UInt8(value & 0xFF))
}

func __pytra_crc32(_ data: [UInt8]) -> UInt32 {
    var crc: UInt32 = 0xFFFFFFFF
    for b in data {
        var x = crc ^ UInt32(b)
        var i = 0
        while i < 8 {
            if (x & 1) != 0 {
                x = (x >> 1) ^ 0xEDB88320
            } else {
                x >>= 1
            }
            i += 1
        }
        crc = x
    }
    return crc ^ 0xFFFFFFFF
}

func __pytra_adler32(_ data: [UInt8]) -> UInt32 {
    var s1: UInt32 = 1
    var s2: UInt32 = 0
    for b in data {
        s1 = (s1 + UInt32(b)) % 65521
        s2 = (s2 + s1) % 65521
    }
    return (s2 << 16) | s1
}

func __pytra_zlib_store(_ data: [UInt8]) -> [UInt8] {
    var out: [UInt8] = [0x78, 0x01]
    var pos = 0
    while pos < data.count {
        let remain = data.count - pos
        let chunkLen = min(65535, remain)
        let isFinal: UInt8 = (pos + chunkLen >= data.count) ? 1 : 0
        out.append(isFinal)

        let len = UInt16(chunkLen)
        out.append(UInt8(len & 0xFF))
        out.append(UInt8((len >> 8) & 0xFF))
        let nlen = ~len
        out.append(UInt8(nlen & 0xFF))
        out.append(UInt8((nlen >> 8) & 0xFF))

        out.append(contentsOf: data[pos..<(pos + chunkLen)])
        pos += chunkLen
    }
    __pytra_append_u32be(&out, __pytra_adler32(data))
    return out
}

func __pytra_png_chunk(_ kind: [UInt8], _ payload: [UInt8]) -> [UInt8] {
    var out: [UInt8] = []
    __pytra_append_u32be(&out, UInt32(payload.count))
    out.append(contentsOf: kind)
    out.append(contentsOf: payload)
    var crcInput = kind
    crcInput.append(contentsOf: payload)
    __pytra_append_u32be(&out, __pytra_crc32(crcInput))
    return out
}

func __pytra_write_rgb_png(_ path: Any?, _ width: Any?, _ height: Any?, _ pixels: Any?) {
    let outPath = __pytra_str(path)
    let w = Int(__pytra_int(width))
    let h = Int(__pytra_int(height))
    if w <= 0 || h <= 0 {
        return
    }

    let raw = __pytra_as_list(pixels)
    let expected = w * h * 3
    if raw.count < expected {
        return
    }

    var scanlines: [UInt8] = []
    scanlines.reserveCapacity(h * (1 + w * 3))
    var idx = 0
    var y = 0
    while y < h {
        scanlines.append(0)
        var x = 0
        while x < w {
            scanlines.append(__pytra_to_u8(raw[idx]))
            scanlines.append(__pytra_to_u8(raw[idx + 1]))
            scanlines.append(__pytra_to_u8(raw[idx + 2]))
            idx += 3
            x += 1
        }
        y += 1
    }

    var ihdr: [UInt8] = []
    __pytra_append_u32be(&ihdr, UInt32(w))
    __pytra_append_u32be(&ihdr, UInt32(h))
    ihdr.append(contentsOf: [8, 2, 0, 0, 0])
    let idat = __pytra_zlib_store(scanlines)

    var png: [UInt8] = [0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A]
    png.append(contentsOf: __pytra_png_chunk([0x49, 0x48, 0x44, 0x52], ihdr))
    png.append(contentsOf: __pytra_png_chunk([0x49, 0x44, 0x41, 0x54], idat))
    png.append(contentsOf: __pytra_png_chunk([0x49, 0x45, 0x4E, 0x44], []))

    let outURL = URL(fileURLWithPath: outPath)
    let parent = outURL.deletingLastPathComponent()
    if parent.path != "" && parent.path != "." {
        try? FileManager.default.createDirectory(at: parent, withIntermediateDirectories: true)
    }
    try? Data(png).write(to: outURL)
}

func __pytra_to_bytes(_ v: Any?) -> [UInt8] {
    if let arr = v as? [UInt8] {
        return arr
    }
    if let arr = v as? [Any] {
        var out: [UInt8] = []
        out.reserveCapacity(arr.count)
        for item in arr {
            out.append(__pytra_to_u8(item))
        }
        return out
    }
    if let data = v as? Data {
        return [UInt8](data)
    }
    if let s = v as? String {
        return [UInt8](s.utf8)
    }
    return []
}

func __pytra_lzw_encode(_ data: [UInt8], _ minCodeSize: Int = 8) -> [UInt8] {
    if data.isEmpty {
        return []
    }
    let clearCode = 1 << minCodeSize
    let endCode = clearCode + 1
    let codeSize = minCodeSize + 1

    var out: [UInt8] = []
    var bitBuffer = 0
    var bitCount = 0

    func emit(_ code: Int) {
        bitBuffer |= (code << bitCount)
        bitCount += codeSize
        while bitCount >= 8 {
            out.append(UInt8(bitBuffer & 0xFF))
            bitBuffer >>= 8
            bitCount -= 8
        }
    }

    emit(clearCode)
    for v in data {
        emit(Int(v))
        emit(clearCode)
    }
    emit(endCode)
    if bitCount > 0 {
        out.append(UInt8(bitBuffer & 0xFF))
    }
    return out
}

func __pytra_append_u16le(_ out: inout [UInt8], _ value: Int) {
    let v = UInt16(truncatingIfNeeded: value)
    out.append(UInt8(v & 0xFF))
    out.append(UInt8((v >> 8) & 0xFF))
}

func __pytra_grayscale_palette() -> [Any] {
    var p: [Any] = []
    p.reserveCapacity(256 * 3)
    var i: Int64 = 0
    while i < 256 {
        p.append(i)
        p.append(i)
        p.append(i)
        i += 1
    }
    return p
}

func __pytra_save_gif(
    _ path: Any?,
    _ width: Any?,
    _ height: Any?,
    _ frames: Any?,
    _ palette: Any?,
    _ delayCs: Any? = Int64(4),
    _ loop: Any? = Int64(0)
) {
    let outPath = __pytra_str(path)
    let w = Int(__pytra_int(width))
    let h = Int(__pytra_int(height))
    if w <= 0 || h <= 0 {
        return
    }

    let pal = __pytra_to_bytes(palette)
    if pal.count != 256 * 3 {
        return
    }

    let frameBytes = w * h
    let dcs = Int(__pytra_int(delayCs))
    let lp = Int(__pytra_int(loop))
    let frs = __pytra_as_list(frames)

    var out: [UInt8] = []
    out.append(contentsOf: [UInt8]("GIF89a".utf8))
    __pytra_append_u16le(&out, w)
    __pytra_append_u16le(&out, h)
    out.append(0xF7)
    out.append(0)
    out.append(0)
    out.append(contentsOf: pal)
    out.append(contentsOf: [0x21, 0xFF, 0x0B])
    out.append(contentsOf: [UInt8]("NETSCAPE2.0".utf8))
    out.append(0x03)
    out.append(0x01)
    __pytra_append_u16le(&out, lp)
    out.append(0x00)

    for frAny in frs {
        let fr = __pytra_to_bytes(frAny)
        if fr.count != frameBytes {
            return
        }
        out.append(contentsOf: [0x21, 0xF9, 0x04, 0x00])
        __pytra_append_u16le(&out, dcs)
        out.append(0x00)
        out.append(0x00)
        out.append(0x2C)
        __pytra_append_u16le(&out, 0)
        __pytra_append_u16le(&out, 0)
        __pytra_append_u16le(&out, w)
        __pytra_append_u16le(&out, h)
        out.append(0x00)
        out.append(0x08)

        let compressed = __pytra_lzw_encode(fr, 8)
        var pos = 0
        while pos < compressed.count {
            let len = min(255, compressed.count - pos)
            out.append(UInt8(len))
            out.append(contentsOf: compressed[pos..<(pos + len)])
            pos += len
        }
        out.append(0x00)
    }
    out.append(0x3B)

    let outURL = URL(fileURLWithPath: outPath)
    let parent = outURL.deletingLastPathComponent()
    if parent.path != "" && parent.path != "." {
        try? FileManager.default.createDirectory(at: parent, withIntermediateDirectories: true)
    }
    try? Data(out).write(to: outURL)
}

func __pytra_bytearray(_ initValue: Any?) -> [Any] {
    if let i = initValue as? Int64 {
        return Array(repeating: Int64(0), count: max(0, Int(i)))
    }
    if let i = initValue as? Int {
        return Array(repeating: Int64(0), count: max(0, i))
    }
    if let arr = initValue as? [Any] {
        return arr
    }
    return []
}

func __pytra_bytes(_ v: Any?) -> [Any] {
    if let arr = v as? [Any] {
        return arr
    }
    return []
}

func __pytra_list_repeat(_ value: Any, _ count: Any?) -> [Any] {
    var out: [Any] = []
    var i: Int64 = 0
    let n = __pytra_int(count)
    while i < n {
        out.append(value)
        i += 1
    }
    return out
}

func __pytra_as_list(_ v: Any?) -> [Any] {
    if let arr = v as? [Any] { return arr }
    return []
}

func __pytra_as_u8_list(_ v: Any?) -> [UInt8] {
    if let arr = v as? [UInt8] { return arr }
    return []
}

func __pytra_as_dict(_ v: Any?) -> [AnyHashable: Any] {
    if let dict = v as? [AnyHashable: Any] { return dict }
    return [:]
}

func __pytra_dict_get(_ dictAny: Any?, _ key: Any?, _ defaultValue: Any?) -> Any {
    if let dict = dictAny as? [AnyHashable: Any] {
        let k = AnyHashable(__pytra_str(key))
        if let v = dict[k] {
            return v
        }
    }
    return defaultValue as Any
}

func __pytra_enumerate(_ v: Any?) -> [Any] {
    let arr = __pytra_as_list(v)
    var out: [Any] = []
    out.reserveCapacity(arr.count)
    var i: Int64 = 0
    while i < Int64(arr.count) {
        out.append([i, arr[Int(i)]])
        i += 1
    }
    return out
}

func __pytra_pop_last(_ v: [Any]) -> [Any] {
    if v.isEmpty { return v }
    return Array(v.dropLast())
}

func __pytra_print(_ args: Any...) {
    if args.isEmpty {
        Swift.print()
        return
    }
    Swift.print(args.map { String(describing: $0) }.joined(separator: " "))
}

func __pytra_min(_ a: Any?, _ b: Any?) -> Any {
    let af = __pytra_float(a)
    let bf = __pytra_float(b)
    if af < bf {
        if __pytra_is_float(a) || __pytra_is_float(b) { return af }
        return __pytra_int(a)
    }
    if __pytra_is_float(a) || __pytra_is_float(b) { return bf }
    return __pytra_int(b)
}

func __pytra_max(_ a: Any?, _ b: Any?) -> Any {
    let af = __pytra_float(a)
    let bf = __pytra_float(b)
    if af > bf {
        if __pytra_is_float(a) || __pytra_is_float(b) { return af }
        return __pytra_int(a)
    }
    if __pytra_is_float(a) || __pytra_is_float(b) { return bf }
    return __pytra_int(b)
}

func __pytra_is_int(_ v: Any?) -> Bool {
    return (v is Int) || (v is Int64)
}

func __pytra_is_float(_ v: Any?) -> Bool {
    return v is Double
}

func __pytra_is_bool(_ v: Any?) -> Bool {
    return v is Bool
}

func __pytra_is_str(_ v: Any?) -> Bool {
    return v is String
}

func __pytra_is_list(_ v: Any?) -> Bool {
    return v is [Any]
}
